dht_net.go

```diff
@@ -266,49 +288,166 @@ func (ms *messageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb
+			return nil, rctx.Err()
+
+		case <-ms.dht.ctx.Done():
+			return nil, ms.dht.ctx.Err()
+		}
+
+		if res.err != nil {
+			ms.lk.Lock()
+			ms.resetSoft(rcount)
+			ms.lk.Unlock()
 
 			if retry {
-				log.Info("error reading message, bailing: ", err)
-				return nil, err
+				log.Info("error reading message, bailing: ", res.err)
+				return nil, res.err
 			} else {
-				log.Info("error reading message, trying again: ", err)
+				log.Info("error reading message, trying again: ", res.err)
 				retry = true
 				continue
 			}
 		}
 
-		log.Event(ctx, "dhtSentMessage", ms.dht.self, ms.p, pmes)
-
-		if ms.singleMes > streamReuseTries {
-			ms.s.Close()
-			ms.s = nil
-		} else if retry {
+		if retry {
+			ms.lk.Lock()
 			ms.singleMes++
+			if ms.singleMes > streamReuseTries {
+				ms.resetSoft(rcount)
+			}
+			ms.lk.Unlock()
 		}
 
-		return mes, nil
+		return res.mes, nil
 	}
 }
 
-func (ms *messageSender) ctxReadMsg(ctx context.Context, mes *pb.Message) error {
+func (ms *messageSender) sendMessageSingle(ctx context.Context, pmes *pb.Message) error {
+	s, err := ms.dht.host.NewStream(ctx, ms.p, ProtocolDHT, ProtocolDHTOld)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	w := ggio.NewDelimitedWriter(s)
+
+	err = w.WriteMsg(pmes)
+	if err != nil {
+		s.Reset()
+	}
+
+	return err
+}
+
+func (ms *messageSender) sendRequestSingle(ctx context.Context, pmes *pb.Message) (*pb.Message, error) {
+	s, err := ms.dht.host.NewStream(ctx, ms.p, ProtocolDHT, ProtocolDHTOld)
+	if err != nil {
+		return nil, err
+	}
+	defer s.Close()
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	if err := w.WriteMsg(pmes); err != nil {
+		s.Reset()
+		return nil, err
+	}
+
+	mes := new(pb.Message)
+
 	errc := make(chan error, 1)
-	go func(r ggio.ReadCloser) {
+	go func() {
 		errc <- r.ReadMsg(mes)
-	}(ms.r)
+	}()
 
-	t := time.NewTimer(dhtReadMessageTimeout)
-	defer t.Stop()
+	rctx, cancel := context.WithTimeout(ctx, dhtReadMessageTimeout)
+	defer cancel()
 
 	select {
 	case err := <-errc:
-		return err
-	case <-ctx.Done():
-		return ctx.Err()
-	case <-t.C:
-		return ErrReadTimeout
+		if err != nil {
+			return nil, err
+		}
+	case <-rctx.Done():
+		return nil, rctx.Err()
+	}
+
+	return mes, nil
+}
+
+// Resets the stream unconditionally; increments the reset count.
+// Mutex must be locked.
+func (ms *messageSender) resetHard() {
+	close(ms.rch)
+	ms.s.Reset()
+	ms.s = nil
+	ms.rcount++
+}
+
+// Resets the stream only if the reset count matches the argument
+// Allows multiple read failures in batched concurrent requests with
+// only a single reset between them.
+// Mutex must be locked.
+func (ms *messageSender) resetSoft(rcount int) {
+	if rcount != ms.rcount {
+		return
+	}
+
+	ms.resetHard()
+}
+
+func messageReceiver(ctx context.Context, rch chan chan requestResult, r ggio.ReadCloser) {
+	for {
+		var next chan requestResult
+		var ok bool
+		select {
+		case next, ok = <-rch:
+			if !ok {
+				return
+			}
```

Comment:
This might be a little cleaner if we move the logic at the bottom of the loop into this case. Would save us from having to pre-declare those variables

Reply:
ok, was just trying to avoid excessive indentation.